src/main/scala/firrtl/ir/IR.scala

```diff
@@ -371,9 +371,45 @@ object GroundType {
 abstract class AggregateType extends Type {
   def mapWidth(f: Width => Width): Type = this
 }
-case class UIntType(width: Width) extends GroundType {
+object UIntType {
+  private val maxCached = 1024
+  private val cache = new Array[UIntType](maxCached + 1)
+  val UnknownWidth = new UIntType(ir.UnknownWidth)
+
+  def apply(width: Width): UIntType = width match {
+    // Should this be combined with the memoization in IntWidth?
+    case IntWidth(w) if 0 <= w && w <= maxCached =>
+      val i = w.toInt
+      var t = cache(i)
+      if (t eq null) {
+        t = new UIntType(width)
+        cache(i) = t
+      }
+      t
+    case ir.UnknownWidth => UnknownWidth
+    case other => new UIntType(width)
+  }
+  // For pattern matching
+  def unapply(u: UIntType): Option[Width] = Some(u.width)
+}
+class UIntType(val width: Width) extends GroundType with Product {
```

Comment:
also add copy.

Actually, if you could just implement the features from the spec....
https://www.scala-lang.org/files/archive/spec/2.12/05-classes-and-objects.html#case-classes